src/dbnode/encoding/series_iterator_accumulator.go

```diff
@@ -0,0 +1,256 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package encoding
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+var _ SeriesIteratorAccumulator = (*seriesIteratorAccumulator)(nil)
+
+type seriesIteratorAccumulator struct {
+	id              ident.ID
+	nsID            ident.ID
+	start           time.Time
+	end             time.Time
+	iters           iterators
+	seriesIterators []SeriesIterator
+	err             error
+	firstNext       bool
+	closed          bool
+}
+
+// NewSeriesIteratorAccumulator creates a new series iterator.
+func NewSeriesIteratorAccumulator(
+	iter SeriesIterator,
+) (SeriesIteratorAccumulator, error) {
+	it := &seriesIteratorAccumulator{
+		id:              iter.ID(),
+		nsID:            iter.Namespace(),
+		seriesIterators: make([]SeriesIterator, 0, 2),
+	}
+
+	err := it.Add(iter)
+	if err != nil {
+		return nil, err
+	}
+
+	return it, nil
+}
+
+func (it *seriesIteratorAccumulator) Add(iter SeriesIterator) error {
+	if newNs := iter.Namespace(); !newNs.Equal(it.nsID) {
+		return fmt.Errorf("cannot add iterator with namespace %s to accumulator %s",
+			newNs.String(), it.nsID.String())
+	}
+
+	if !iter.Next() || !it.iters.push(iter) {
+		iter.Close()
+		return iter.Err()
+	}
+
+	iterStart := iter.Start()
+	if start := it.start; start.IsZero() || iterStart.Before(start) {
+		it.start = iterStart
+	}
+
+	iterEnd := iter.End()
+	if end := it.end; end.IsZero() || iterEnd.After(end) {
+		it.end = iterEnd
+	}
+
+	it.seriesIterators = append(it.seriesIterators, iter)
+	return nil
+
+}
+
+func (it *seriesIteratorAccumulator) ID() ident.ID {
+	return it.id
+}
+
+func (it *seriesIteratorAccumulator) Namespace() ident.ID {
+	return it.nsID
+}
+
+func (it *seriesIteratorAccumulator) Tags() ident.TagIterator {
+	it.err = errors.New("cannot get tags from series iterator accumulator")
+	return ident.EmptyTagIterator
+}
+
+func (it *seriesIteratorAccumulator) Start() time.Time {
+	return it.start
+}
+
+func (it *seriesIteratorAccumulator) End() time.Time {
+	return it.end
+}
+
+func (it *seriesIteratorAccumulator) Next() bool {
+	if !it.firstNext {
+		if !it.hasNext() {
+			return false
+		}
+
+		it.moveToNext()
+	}
+
+	it.firstNext = false
+	return it.hasNext()
+}
+
+func (it *seriesIteratorAccumulator) Current() (ts.Datapoint, xtime.Unit, ts.Annotation) {
+	return it.iters.current()
+}
+
+func (it *seriesIteratorAccumulator) Err() error {
+	if it.err != nil {
+		return it.err
+	}
+
+	for _, iter := range it.seriesIterators {
+		if err := iter.Err(); err != nil {
+			it.err = err
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (it *seriesIteratorAccumulator) Replicas() []MultiReaderIterator {
```

Comment:
Same here maybe we just remove the "Replicas()" call from "SeriesIteratorAccumulator"?